File: week12/19/P3.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    vector<vector<int>> dfs_util(int v, int w, int n_row, int n_col,
                                 vector<vector<char>>& board,
                                 vector<vector<int>>& visited) {
        visited[v][w] = 1;
        if (0 <= v-1 && not visited[v-1][w] && board[v-1][w] == 'X')
            visited = dfs_util(v-1, w, n_row, n_col, board, visited);
        if (v+1 < n_row && not visited[v+1][w] && board[v+1][w] == 'X')
            visited = dfs_util(v+1, w, n_row, n_col, board, visited);
        if (0 <= w-1 && not visited[v][w-1] && board[v][w-1] == 'X')
            visited = dfs_util(v, w-1, n_row, n_col, board, visited);
        if (w+1 < n_col && not visited[v][w+1] && board[v][w+1] == 'X')
            visited = dfs_util(v, w+1, n_row, n_col, board, visited);
        return visited;
    }

    int countBattleships(vector<vector<char>>& board) {
        int n_row = board.size();
        int n_col = board[0].size();
        vector<vector<int>> visited = {};
        for (int i=0; i<n_row; i++) {
            vector<int> temp(n_col, 0);
            visited.push_back(temp);
        }

        int cnt = 0;
        for (int i=0; i<n_row; i++) {
            for (int j=0; j<n_col; j++) {
                if (visited[i][j] == 0 && board[i][j] == 'X') {
                    visited = dfs_util(i, j, n_row, n_col, board, visited);
                    cnt++;
                }
            }
        }

        return cnt;
    }
};


int main() {
    vector<vector<char>> board = {{'X','.','.','X'},{'.','.','.','X'},{'.','.','.','X'}};
    Solution result;
    cout << result.countBattleships(board) << endl;

    return 0;
}

File: week12/19/P3.py
class Solution(object):
    def countBattleships(self, board):
        n_row = len(board)
        n_col = len(board[0])
        visited = [[False for _ in range(n_col)] for _ in range(n_row)]
        
        def dfs_util(v, w):
            visited[v][w] = True
            if (0 <= v-1 < n_row and 0 <= w < n_col and \
                not visited[v-1][w] and board[v-1][w] == "X"):
                dfs_util(v-1, w)
            if (0 <= v+1 < n_row and 0 <= w < n_col and \
                not visited[v+1][w] and board[v+1][w] == "X"):
                dfs_util(v+1, w)
            if (0 <= v < n_row and 0 <= w-1 < n_col and \
                not visited[v][w-1] and board[v][w-1] == "X"):
                dfs_util(v, w-1)
            if (0 <= v < n_row and 0 <= w+1 < n_col and \
                not visited[v][w+1] and board[v][w+1] == "X"):
                dfs_util(v, w+1)

        cnt = 0
        for i in range(n_row):
            for j in range(n_col):
                if not visited[i][j] and board[i][j] == "X":
                    dfs_util(i, j)
                    cnt += 1        

        return cnt

if __name__ == "__main__":
    result = Solution()
    board = [["X",".",".","X"],[".",".",".","X"],[".",".",".","X"]]
    print(result.countBattleships(board))
    board = [["."]]
    print(result.countBattleships(board))